Add a note to documentation about toolchains limitations with musl library (#24511)
gradle/gradle: subprojects/docs/src/docs/userguide/jvm/toolchains.adoc

This can be achieved by passing an empty lambda when requesting a tool: `javaToolchainService.launcherFor({})`.

You can find more details on defining custom tasks in the <<more_about_tasks.adoc#more_about_tasks, Authoring tasks>> documentation.

[[sec:limitations]]
== Toolchains limitations

> Gradle may detect toolchains incorrectly when a JVM/JRE compiled with `musl` library is used.

Reviewer: I think we should also include a brief, one-line description of what the `musl` library is.  Something like "The musl library is an alternative implementation of the C standard library focused on standards-conformance."

> Gradle may detect toolchains incorrectly when the build is run in a JVM/JRE compiled against `musl`, https://musl.libc.org/[an alternative implementation] of the C standard library.

Reviewer 2: ```suggestion
Gradle may detect toolchains incorrectly when it's running in a JVM compiled against `musl`, an https://musl.libc.org/[alternative implementation] of the C standard library.
```

> A JVM/JRE compiled against `musl` may override the `LD_LIBRARY_PATH` environment variable in order to control dynamic library resolution.

Reviewer 2: ```suggestion
JVMs compiled against `musl` can sometimes override the `LD_LIBRARY_PATH` environment variable to control dynamic library resolution.
```

> This might affect java subprocesses launched by Gradle, resulting in unexpected behavior.

Reviewer 2: ```suggestion
This can influence forked java processes launched by Gradle for probing toolchains, resulting in unexpected behavior.
```

Author: It's not just toolchain probing, any java subprocess is affected.

This might affect java subprocesses launched by Gradle, resulting in unexpected behavior.

> Thus, usage of multiple java toolchains is discouraged in environments with `musl` library.

Reviewer: Is there a situation where toolchains will work when running in a `musl` JVM?  For instance, if I always specify toolchains with the same version as the JVM the build is running in, will that work?  It's not clear what distinction we are making with "multiple toolchains" here (i.e. does that mean it works if I have a single toolchain?)  We should either say "don't use toolchains with `musl`" or we should say "toolchains will work in a very specific scenario" and describe what that scenario is.  We should be clear about what we are suggesting (I think it's totally ok to say "just don't use toolchains" even though it might work in certain circumstances).

Author: First of all, thanks for review!

IIUC, Gradle would work fine if a single toolchain is used. I think this might be qualified as a pretty common scenario. In my experience, most companies don't care about multiple JVM versions support because their software would only be run in cloud/on-premise servers.

> Thus, usage of multiple java toolchains is discouraged in environments with the `musl` library.

Reviewer 2: ```suggestion
As a consequence, using multiple different toolchains in a build is discouraged in environments based on the `musl` library.
```

Reviewer 2: This is not clear enough. What exactly is the non-recommended scenario? Running Gradle with one Java version, having a toolchain configured with another version? Having two different toolchains configured in the build with two Java versions? Please clarify, so that whoever reads this knows exactly what's ok to do and what not.

Author: `environments based on` doesn't sound right to me. Ubuntu is not based on musl, but you can have a JVM compiled against musl there.

The same goes for `in build` – it's not just building, any java subprocess can be affected.  

As for clarity - I guess the last sentence clears this, doesn't it?
`If you are using a single toolchain to build and run your application, you can safely ignore this limitation.`

> For example, this applies to most `alpine` containers — consider using another OS, e.g. `ubuntu`.

Reviewer: ```suggestion
For example, this applies to most `alpine` distributions — consider using another distribution, e.g. `ubuntu`.
```

> For example, this applies to most `alpine` distributions — consider using another distribution, e.g. `ubuntu`.

Reviewer 2: ```suggestion
This is the case in most `Alpine` distributions — consider using another distribution, like Ubuntu, instead.
```

> If you are using a single toolchain to build and run your application, you can safely ignore this limitation.

Reviewer 2: ```suggestion
If you are using a single toolchain, the JVM running Gradle, to build and run your application, you can safely ignore this limitation.
```
